Modules/_decimal/_decimal.c

```diff
@@ -2323,38 +2322,42 @@ static PyObject *
 dec_from_long(decimal_state *state, PyTypeObject *type, PyObject *v,
               const mpd_context_t *ctx, uint32_t *status)
 {
-    PyObject *dec;
-    PyLongObject *l = (PyLongObject *)v;
+    PyObject *dec = PyDecType_New(state, type);
 
-    dec = PyDecType_New(state, type);
     if (dec == NULL) {
         return NULL;
     }
 
-    if (_PyLong_IsZero(l)) {
-        _dec_settriple(dec, MPD_POS, 0, 0);
-        return dec;
-    }
-
-    uint8_t sign = _PyLong_IsNegative(l) ? MPD_NEG :  MPD_POS;
+    PyLongExport export_long;
 
-    if (_PyLong_IsCompact(l)) {
-        _dec_settriple(dec, sign, l->long_value.ob_digit[0], 0);
-        mpd_qfinalize(MPD(dec), ctx, status);
-        return dec;
+    if (PyLong_Export(v, &export_long) == -1) {
+        Py_DECREF(dec);
+        return NULL;
     }
-    size_t len = _PyLong_DigitCount(l);
-
-#if PYLONG_BITS_IN_DIGIT == 30
-    mpd_qimport_u32(MPD(dec), l->long_value.ob_digit, len, sign, PyLong_BASE,
-                    ctx, status);
-#elif PYLONG_BITS_IN_DIGIT == 15
-    mpd_qimport_u16(MPD(dec), l->long_value.ob_digit, len, sign, PyLong_BASE,
-                    ctx, status);
-#else
-  #error "PYLONG_BITS_IN_DIGIT should be 15 or 30"
-#endif
+    if (export_long.digits) {
+        const PyLongLayout *layout = PyLong_GetNativeLayout();
+        const uint8_t bpd = layout->bits_per_digit;
+        const uint8_t sign = export_long.negative ? MPD_NEG :  MPD_POS;
+        const Py_ssize_t len = export_long.ndigits;
 
+        if (bpd == 30) {
+            mpd_qimport_u32(MPD(dec), export_long.digits, len, sign,
+                            1 << bpd, ctx, status);
+        }
+        else if (bpd == 15) {
+            mpd_qimport_u16(MPD(dec), export_long.digits, len, sign,
+                            1 << bpd, ctx, status);
+        }
+        else {
+            PyLong_FreeExport(&export_long);
+            Py_DECREF(dec);
+            return NULL;
```

Comment:
You must raise an exception in this case. You can raise a RuntimeError("bits_per_digit must be 15 or 30").

Reply:
Yeah, I've decided to use ``#if``'s again.  We can query for native layout, cache that in the module state, but it hardly does make sense for a stdlib module.